File: server/app/schemas/scheduler.py
from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field


class BackupEngine(StrEnum):
    FILESYSTEM = "filesystem"
    MARIADB = "mariadb"


class BackupJobStatus(StrEnum):
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    RETRYING = "retrying"


class BackupPolicyCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    repository_uuid: str = Field(min_length=36, max_length=36)
    engine: BackupEngine
    source: str = Field(min_length=1, max_length=4096)
    interval_minutes: int = Field(ge=1, le=525600)
    enabled: bool = True
    next_run_at: datetime
    max_retries: int = Field(default=3, ge=0, le=10)


class BackupPolicyUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    source: str | None = Field(default=None, min_length=1, max_length=4096)
    interval_minutes: int | None = Field(default=None, ge=1, le=525600)
    enabled: bool | None = None
    next_run_at: datetime | None = None
    max_retries: int | None = Field(default=None, ge=0, le=10)


class BackupPolicyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    uuid: str
    repository_uuid: str
    name: str
    engine: BackupEngine
    source: str
    interval_minutes: int
    enabled: bool
    next_run_at: datetime
    max_retries: int
    created_at: datetime
    updated_at: datetime


class BackupJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    uuid: str
    policy_uuid: str
    status: BackupJobStatus
    attempts: int
    max_attempts: int
    queued_at: datetime
    started_at: datetime | None
    completed_at: datetime | None
    error_message: str | None
    snapshot_uuid: str | None
    created_at: datetime
    updated_at: datetime


class SchedulerRunResponse(BaseModel):
    enqueued_jobs: list[BackupJobResponse]


class WorkerRunResponse(BaseModel):
    job: BackupJobResponse | None

File: frontend/src/stores/scheduler.ts
import { defineStore } from 'pinia'

import api from '../api'

export type BackupEngine = 'filesystem' | 'mariadb'
export type BackupJobStatus = 'queued' | 'running' | 'completed' | 'failed' | 'retrying'

export interface BackupPolicy {
  uuid: string
  repository_uuid: string
  name: string
  engine: BackupEngine
  source: string
  interval_minutes: number
  enabled: boolean
  next_run_at: string
  max_retries: number
  created_at: string
  updated_at: string
}

export interface BackupJob {
  uuid: string
  policy_uuid: string
  status: BackupJobStatus
  attempts: number
  max_attempts: number
  queued_at: string
  started_at: string | null
  completed_at: string | null
  error_message: string | null
  snapshot_uuid: string | null
  created_at: string
  updated_at: string
}

export interface BackupPolicyForm {
  name: string
  repositoryUuid: string
  engine: BackupEngine
  source: string
  intervalMinutes: number
  nextRunAt: string
  maxRetries: number
}

interface SchedulerState {
  policies: BackupPolicy[]
  jobs: BackupJob[]
  loading: boolean
  saving: boolean
  error: string | null
}

export const useSchedulerStore = defineStore('scheduler', {
  state: (): SchedulerState => ({
    policies: [],
    jobs: [],
    loading: false,
    saving: false,
    error: null,
  }),
  actions: {
    async loadScheduler(): Promise<void> {
      this.loading = true
      this.error = null
      try {
        const [policiesResponse, jobsResponse] = await Promise.all([
          api.get<BackupPolicy[]>('/scheduler/policies'),
          api.get<BackupJob[]>('/scheduler/jobs'),
        ])
        this.policies = policiesResponse.data
        this.jobs = jobsResponse.data
      } catch (error) {
        this.error = 'Scheduler state could not be loaded.'
      } finally {
        this.loading = false
      }
    },
    async createPolicy(form: BackupPolicyForm): Promise<void> {
      this.saving = true
      this.error = null
      try {
        const response = await api.post<BackupPolicy>('/scheduler/policies', {
          name: form.name,
          repository_uuid: form.repositoryUuid,
          engine: form.engine,
          source: form.source,
          interval_minutes: form.intervalMinutes,
          enabled: true,
          next_run_at: new Date(form.nextRunAt).toISOString(),
          max_retries: form.maxRetries,
        })
        this.policies = [...this.policies, response.data].sort((left, right) =>
          left.name.localeCompare(right.name),
        )
      } catch (error) {
        this.error = 'Schedule could not be created.'
        throw error
      } finally {
        this.saving = false
      }
    },
    async enqueuePolicy(policyUuid: string): Promise<void> {
      this.error = null
      try {
        const response = await api.post<BackupJob>(
          `/scheduler/policies/${policyUuid}/enqueue`,
        )
        this.jobs = [response.data, ...this.jobs]
      } catch (error) {
        this.error = 'Schedule could not be enqueued.'
        throw error
      }
    },
    async runWorker(): Promise<void> {
      this.error = null
      try {
        await api.post('/scheduler/work')
        await this.loadScheduler()
      } catch (error) {
        this.error = 'Worker could not process a job.'
        throw error
      }
    },
  },
})
